packages/@aws-cdk/core/lib/names.ts

```diff
@@ -36,5 +39,26 @@ export class Names {
     return components.length > 0 ? makeUniqueId(components) : '';
   }
 
+  /**
+   * Returns a CloudFormation-compatible unique identifier for a construct based
+   * on its path. If the node is a Stack, the stack name is used for that path component instead.
+   * The identifier includes a human readable portion rendered from the path components and a hash suffix.
+   * @param construct The construct
+   * @param maxUniqueIdLength The maximum length of the unique id
+   * @returns a unique id based on the construct path
+   */
+  public static stackSafeUniqueId(construct: Construct, maxUniqueIdLength: number = 256): string {
+    const node = Node.of(construct);
+    const components = node.scopes.slice(1).map(component => {
+      if (Stack.isStack(component) && !unresolved(component.stackName)) {
+        return component.stackName;
+      }
+      return Node.of(component).id;
```

Comment:
Any chance of writing this as a single statement instead?

```suggestion
      return Stack.isStack(component) && !unresolved(component.stackName)
        ? component.stackName
        : Node.of(component).id;
```

Reply:
I removed this particular piece of code and went in another direction per comments above.